src/main/java/org/spongepowered/api/data/key/KeyFactory.java

```diff
@@ -196,53 +196,37 @@ public String toString() {
      * @return The generated key
      */
-    public static <E> Key<SetValue<E>> makeSetKey(final TypeToken<? extends Set<E>> elementToken, TypeToken<SetValue<E>> valueToken,
-            final DataQuery query, final String id, final String name) {
-        validateId(id);
-        return new Key<SetValue<E>>() {
-
-            @Nullable private String string;
-
-            @Override
-            public String getId() {
-                return id;
-            }
-
-            @Override
-            public String getName() {
-                return name;
-            }
-
-            private final int hash = Objects.hashCode(ListValue.class, elementToken, query);
-
-            @Override
-            public TypeToken<SetValue<E>> getValueToken() {
-                return valueToken;
-            }
-
-            @Override
-            public TypeToken<?> getElementToken() {
-                return elementToken;
-            }
-
-            @Override
-            public DataQuery getQuery() {
-                return query;
-            }
-
-            @Override
-            public int hashCode() {
-                return this.hash;
-            }
+    public static <E> Key<SetValue<E>> makeSetKey(TypeToken<? extends Set<E>> elementToken, TypeToken<SetValue<E>> valueToken,
+            DataQuery query, String id, String name) {
+        return makeSingleKey((TypeToken<Set<E>>) elementToken, valueToken, query, id, name);
+    }
 
-            @Override
-            public String toString() {
-                if (this.string == null) {
-                    this.string = "Key{Value:" + "SetValue<" + elementToken.toString() + ">, Query: " + query.toString() + "}";
-                }
-                return this.string;
-            }
-        };
+    /**
+     * Creates a new {@link Key} based on a {@link MapValue} of the types
+     * <code>K</code> keys and <code>V</code> values with the provided
+     * {@link DataQuery} for accessing the {@link Map} in {@link DataView}s.
+     * <p>This is a shortcut for {@link #makeMapKey(TypeToken, TypeToken, DataQuery, String, String)}
+     * where you only need to specify the {@link TypeToken}s for the key
+     * {@link K} and value {@link V}.</p>
+     *
+     * @param keyToken The type token of the map keys
+     * @param valueToken The type token of the map values
+     * @param query The query
+     * @param id The id for the new key
+     * @param name The name for the new key
+     * @param <K> The type of keys
+     * @param <V> The type of values
+     * @return The generated key
+     */
+    public static <K, V> Key<MapValue<K, V>> makeMapKeyWithKeyAndValue(TypeToken<K> keyToken, TypeToken<V> valueToken,
+            DataQuery query, String id, String name) {
+        final TypeToken<Map<K, V>> elementToken = new TypeToken<Map<K, V>>() {}
+                .where(new TypeParameter<K>() {}, keyToken)
+                .where(new TypeParameter<V>() {}, valueToken);
```

Comment:
This won’t save your generics as they were initially intended. It’s going to get raw generics at runtime and compile time due to the lack of the creation at the compilation level.